### DISCUSSION TITLE
GSoC 2025 Proposal 4: DeepForest Vision Agent connection with LandingAI

### DISCUSSION THREAD

**Discussion by Samia35-2973** [weecology/DeepForest#973] [General]
Hi _Henry_, _Ben_, and _Ethan_,

I am _Samia Haque_, a recent graduate with a B.Sc. in Software Engineering (major in Data Science). I have been actively working on integrating computer vision with post-hoc interpretability techniques and LLM-based automation, exploring how these technologies can enhance decision-making in various domains.

### My Skills and Experience

- **Object Detection & Post-hoc AI Explainability:**  
  I am currently working on multi-stage traffic anomaly detection, where I utilized YOLOv9 and YOLOv10 to detect traffic congested regions from images. I am also working on its post-hoc interpretability and LLM integration.

- **LLM Integration & AI Automation:**  
  I have experience in prompt engineering, dataset curation, and human feedback-driven improvements for LLMs. While I am not deeply involved in LLM fine-tuning or RLHF at a model-training level, my work in structuring datasets and evaluating LLM outputs aligns with Vision Agent’s AI-driven image analysis pipeline.

I am excited to contribute to GSoC 2025, particularly through **Proposal 4: DeepForest Vision Agent connection with LandingAI** because it combines object detection with AI-driven automation. The idea of integrating a Vision Agent for automated labeling and active learning complements my ongoing exploration of AI-driven decision-making systems. Currently, I am studying the DeepForest codebase, documentation, and related research papers to better understand its architecture.

I would appreciate guidance on recommended resources or discussions for better understanding Vision Agent’s integration.

_Thanks,_  
_Samia_

**Contact Details:**
- GitHub: [Samia35-2973](https://github.com/Samia35-2973)
- LinkedIn: [Samia Haque](https://www.linkedin.com/in/samiahaquetisha/)
- Email: samiatisha35@gmail.com
- Discord Username: _samia_tisha_

**Comment by henrykironde** [comment]
Hi @Samia35-2973,  Thank you for reaching out and for your interest in contributing to DeepForest through GSoC 2025! 

> https://github.com/landing-ai/vision-agent?tab=readme-ov-file
> 
> Rationale:
> 
> Text-based queries of images for labeling and organization.
> 
> Approach:
> 
> Create configuration for DeepForest users to register LLM keys
> Object detection and segmentation workflows
> Develop a user-friendly interface for selecting new images based on agent responses
> Evaluate the effectiveness of the active learning module in improving model accuracy.
> Expected Outcomes:
> 
> An agent-interaction module for DeepForest using VisionAgent. Documentation on using the agent module.

Base on the project details as  above, 
To get started, I recommend thoroughly reviewing LandingAI’s Vision Agent: [LandingAI Vision Agent GitHub](https://github.com/landing-ai/vision-agent?tab=readme-ov-file)

Here is an example of the demo from the landing-ai service given a text query of `hat` shows the detections as shown below.

<img width="1502" alt="Screenshot 2025-03-14 at 12 51 50 AM" src="https://github.com/user-attachments/assets/a331e7e1-b1ee-4a0c-9b71-403184fa0f36" />


Try out all DeepForest models and their detection classes to understand how they could be leveraged for a user-friendly interface that selects new images based on agent responses.

Think about how we can design a user-friendly interface that allows users to input text-based queries, load a model, and detect objects in images based on the query.

**Reply by Samia35-2973** [reply]
Thanks for this helpful guide @henrykironde. I'll start exploring LandingAI’s Vision Agent and the DeepForest model testing ASAP. I'll keep you updated as I make progress.

**Comment by bw4sz** [comment]
Thanks for your interest. I will say that this is more exploratory than the average GSOC proposal, but I think has great potential. There are huge amounts of historical imagery that were annotated in ways that prevent machine learning development, things like place markers and dots over the images. I think connecting the LLMs and Deepforest together to generate code or perform agent-like tasks has alot of potential. Here is a sample idea to play with, i'll post an image when the authors let me. 

I think one place that LLMs can start to play a role is in resurrecting historically annotated data. There are huge datasets from prior research efforts pre2020 that were not annotated with any goal of machine learning development. Consider this image in which a non-profit directly 'dotted' the image, literally overtop the image pixel, obscuring the target classes. We have the original image, the counts, and a 'screenshot' of the final dotted image. I suspect that an object detector (like our bird detector) plus a LLM may be able to recover the positions of these annotations and convert them into COCO like annotations. 

<img width="1435" alt="Screenshot 2025-03-17 at 2 26 24 PM" src="https://github.com/user-attachments/assets/08c4ad63-2433-41dc-9565-656a9b559b75" />

<img width="1763" alt="image" src="https://github.com/user-attachments/assets/30c368fd-be36-433f-9a19-fafe18bf459e" />

**Reply by Samia35-2973** [reply]
Thanks @bw4sz, for sharing this idea. I find the challenge of recovering annotation positions and converting them into COCO format from historically annotated data really interesting. I'm currently exploring the Vision Agent, working with images and queries, and I would really appreciate the sample image when it becomes available so I can experiment with this idea. Are there any specific datasets or past research papers you'd recommend reviewing for similar challenges? I’m also curious about any potential challenges you foresee in aligning historical annotations with modern ML-compatible formats.

**Comment by bw4sz** [comment]
This particular case is just an example, but yes I will pass you an image. At a higher level, we want some connection between DeepForest predictions, images, and LLMs with API keys. It may be independent of vision agent, the project name is somewhat wrong, it could be more like "Connecting ecological computer vision with large language models".Here is a different example from today, a user has 100,000 images, and runs each image through the bird detector. We want to then ask a LLM, like claude or chatgpt, or vision agent, to describe the relationships in the image. I am imagining the user provides a API key and we engineer the infrastructure to apply it to images and detections. 
<img width="1508" alt="image" src="https://github.com/user-attachments/assets/328714ce-491d-4dbc-a7eb-cfbdb7837764" />

**Reply by Samia35-2973** [reply]
> It may be independent of vision agent, the project name is somewhat wrong, it could be more like "Connecting ecological computer vision with large language models"

I see! Thanks for the clarification! That makes a lot of sense.

If I understand correctly, the goal is to build an infrastructure where users provide an API key, DeepForest runs detections on images, and an LLM interprets the relationships between objects or generates insights based on user queries. Do you see this working as a general pipeline for multiple LLMs, or should we focus on integrating with a specific one first?

**Reply by bw4sz** [reply]
Always start simple, choose any LLM we want and we can generalize later. I have been reading https://arxiv.org/pdf/2501.02189. I don’t know about vpn access but our university has set up API endpoints for many of the common LLMs. I’d have to see about outside access.

**Reply by Samia35-2973** [reply]
I went through the paper and started experimenting with LLaMA 3.2 - 90B Vision alongside the DeepForest Bird Detector Model. I converted the detection output of the Bird Detector model into a structured query and combined it with the corresponding image to generate a response from LLaMA. I’ve encountered some challenges like downloading the model takes a significant amount of time, and in some cases, the process times out before completion. I’m still working on troubleshooting this, and if I’m able to run it successfully, I’ll share a snapshot here.
I’d like to confirm if I’m on the right track up to the response generation step. Does this align with the work expected in the proposal? I understand that we also need to build the user interface and implement active learning. I’m currently trying to deepen my understanding of active learning mechanisms and have been reviewing this paper: https://arxiv.org/html/2410.22187v1. Please let me know if there are specific aspects of the paper that you’d like me to focus on. 
Additionally, should we explore integrating unique capabilities as you mentioned in the examples previously, such as annotation recovery from historical ecological data or forest change detection by comparing DeepForest detections on timely versions of the same dataset (e.g., detecting changes in tree coverage or species count by analyzing past vs. present images of the same region)?

**Reply by bw4sz** [reply]
Separate out the active learning part, that's too much for one proposal, I can update the wiki. If we can get a clean connection between a deepforest model detections, and a LLM that can incorporate prompts and generate natural language, that will be a huge step.

**Reply by Samia35-2973** [reply]
Got it! I’ll start drafting my proposal based on this approach and would love to get feedback from you all. What would be the best way to share it? should I post it here for discussion, or send it via email or another method?

**Reply by henrykironde** [reply]
The GSoC dashboard is the right place, as long as you keep it in draft mode until it's ready for submission.

**Reply by bw4sz** [reply]
can you drop the link here, or @henrykironde remind me how to see it? This proposal needs more feedback than the others.

**Reply by Samia35-2973** [reply]
Hi @bw4sz and @henrykironde,
I’ve drafted a proposal for integrating DeepForest’s object detection models with LLMs for ecological image analysis and querying ([link to proposal](https://docs.google.com/document/d/1mubuc_JuISeFWdWeUEvvmmybyHFbdlnmVJD0J_i5PKk/edit?usp=sharing)). 
While I couldn’t experiment with Llama due to hardware limitations, I built a prototype using Gemini API and DeepForest’s bird detection model. I used DeepForest’s annotations from a user-provided image, added some context, and structured them into a prompt to build a sample agent. As shown in this [video](https://drive.google.com/file/d/1iD1q9CV1_fDkEh9lfppJnzr7st43b-Mh/view?usp=sharing), I provided a query and received a response that gives a basic idea of how the system will function. Please ignore the UI quality for now, as the main focus was to check whether the agent responds correctly. There’s still a lot to refine, and I plan to complete this prototype during this phase. 
Please review my proposal and the prototype video, and let me know if anything needs refinement. I’d greatly appreciate your guidance on:
- **Agent Focus Areas**: Are there specific ecological query types (e.g., spatial relationships, species interactions) you’d prioritize for response tailoring?
- **LLM Preferences**: Do you have any preferences for fine-tuning techniques or a preferred LLM, such as Vision-Agent?
- Additionally, are there any UI requirements you’d like me to consider?

**Reply by Samia35-2973** [reply]
I’m trying to upload my proposal as a draft to the GSoC dashboard, and I need to specify the project size. Could you please let me know what I should input there?

**Reply by ethanwhite** [reply]
You can choose the size that matches your availability to do the work. If you have the time to conduct a longer project it's generally a good idea to propose a longer project.

**Reply by Samia35-2973** [reply]
Thanks, Ethan! I chose the larger project size and have also uploaded the draft to the GSoC dashboard.

**Reply by bw4sz** [reply]
No preferences on specific LLM agent, the world is moving so quickly that its nearly impossible to keep track of the options and how they connect it. Choose one that serves as a reasonable starting place and we can innovate from there.

**Reply by Samia35-2973** [reply]
I see. Then, I’ll begin with Gemini as the initial LLM agent since I’ve already built a prototype with it. As the project progresses, I’ll also explore other agents, especially LandingAI’s Vision-Agent to evaluate their fit.

**Comment by Samia35-2973** [comment]
Hi. I've added a basic integration of the Gemini model with the DeepForest tool in this pull request: https://github.com/weecology/deepforest-agent/pull/2. The goal is to test whether Gemini can effectively interact with and call DeepForest tools. I've also included the current workflow diagram in the README file. Could you please run and test it out?

**Reply by bw4sz** [reply]
Very exciting. On my to do list. Great stuff so far!

**Reply by Samia35-2973** [reply]
I just made some changes on this https://github.com/weecology/deepforest-agent/pull/2 pull request. Now it seems to be working better than before. I removed the `predict_image` method in favor of always using `predict_tile` along with the `extension` and `return_plot` parameters. I also fixed the original image handling by appending it properly in messages for Gemini, and improved the system prompt for better responses. The Workflow diagram is changed accordingly in the README file. Please pull the latest changes and run it again.
